### ISSUE TITLE
SystemTextJsonSamples.Utf8ReaderPartialRead is... what is is.

### ISSUE THREAD

**Issue by olivier-spinelli** [dotnet/docs#32950]
The SystemTextJsonSamples.Utf8ReaderPartialRead sample is totall wrong.
The function that SHOULD do the job with the reader state is not even called (and if it was, it will not be able to do anything).

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 73a13d38-4904-c068-c9eb-e43176ce5338
* Version Independent ID: c5dfa4d2-c77a-e752-dae7-0bfae5756b76
* Content: [How to use a JSON document, Utf8JsonReader, and Utf8JsonWriter in System.Text.Json](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter)
* Content Source: [docs/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter.md](https://github.com/dotnet/docs/blob/main/docs/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

---
[Associated WorkItem - 97029](https://dev.azure.com/msft-skilling/Content/_workitems/edit/97029)

**Comment by olivier-spinelli** [comment]
Since I needed it, I wrote a small (but not totally trivial) helper class that does the job:
https://gist.github.com/olivier-spinelli/7785f4f2af33aa8193729f6abf9ee17a

To use it, the reader functions are like `void ReadSomething( ref Utf8JsonReader r, Utf8JsonStreamReader streamReader )`.
And inside, all calls to `r.Read()` must be changed to `if( !r.Read() ) streamReader.ReadMoreData( ref r );` and all calls to `r.Skip()` must be changed to `if( !r.TrySkip() ) streamReader.ReadMoreData( ref r );`.

There are tests for this, just ask if youy want them or if you need me to remove the minor dependencies to my libraries (the `Throw` helper).


If this can help to obtain a better documentation, I'd be happy.

**Comment by gewarren** [comment]
I get a JsonReaderException when I run the sample at https://github.com/dotnet/docs/blob/main/docs/standard/serialization/system-text-json/snippets/how-to/csharp/Utf8ReaderPartialRead.cs with the buffer size set to 20 (line 23). @tdykstra or @eiriktsarpalis can you help me fix it?

```
String in buffer is: {
                "
String in buffer is: "Date": "2019-08-01T
String in buffer is: "2019-08-01T00:00:00
Increased buffer size to 40
String in buffer is: "2019-08-01T00:00:00-07:00",

String in buffer is:       "Temperature":
Unhandled exception. System.Text.Json.JsonReaderException: '"' is invalid after a value. Expected either ',', '}', or ']'. LineNumber: 1 | BytePositionInLine: 57.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeNextToken(Byte marker)
   at System.Text.Json.Utf8JsonReader.ConsumeNextTokenOrRollback(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at SystemTextJsonSamples.Utf8ReaderPartialRead.Main()
```

**Comment by tdykstra** [comment]
@gewarren 

It used to work but [this commit](https://github.com/dotnet/docs/commit/96465874550dffd50bac3495f69f5b3674a99669) removed "ref" from the references to the buffer, and that's what causes it to not work now. Add back ref before buffer in the three places it was removed, and the sample works.

**Comment by gewarren** [comment]
Fixed by #35883
